lib/modules/watcher/index.js

```diff
@@ -0,0 +1,180 @@
+let chokidar = require('chokidar');
+let path = require('path');
+
+let fs = require('../../core/fs.js');
+
+// TODO: this should be receiving the config object not re-reading the
+// embark.json file
+class Watcher {
+  constructor(embark) {
+    this.logger = embark.logger;
+    this.events = embark.events;
+    this.fileWatchers = [];
+
+    this.events.setCommandHandler('watcher:start', () => this.start());
+    this.events.setCommandHandler('watcher:stop', () => this.stop());
+    this.events.setCommandHandler('watcher:restart', () => this.restart());
+  }
+
+  start() {
+    let self = this;
+    // TODO: should come from the config object instead of reading the file
+    // directly
+    let embarkConfig = fs.readJSONSync("embark.json");
+
+    this.watchAssets(embarkConfig, function () {
+      self.logger.trace('ready to watch asset changes');
+    });
+
+    this.watchContracts(embarkConfig, function () {
+      self.logger.trace('ready to watch contract changes');
+    });
+
+    this.watchContractConfig(embarkConfig, function () {
+      self.logger.trace('ready to watch contract config changes');
+    });
+
+    this.watchWebserverConfig(embarkConfig, function () {
+      self.logger.trace('ready to watch webserver config changes');
+    });
+
+    this.logger.info(__("ready to watch file changes"));
+  }
+
+  restart() {
+    this.stop();
+    this.start();
+  }
+
+  stop() {
+    this.fileWatchers.forEach(fileWatcher => {
+      if (fileWatcher.shouldClose) return;
+      if (fileWatcher.isReady) fileWatcher.close();
+      fileWatcher.shouldClose = true;
+    });
+  }
+
+  watchAssets(embarkConfig, callback) {
+    let self = this;
+    let appConfig = embarkConfig.app;
+    let filesToWatch = [];
+
+    for (let targetFile in appConfig) {
+      let files = appConfig[targetFile];
+      let fileGlob = files;
+
+      // workaround for imports issue
+      // so embark reacts to changes made in imported js files
+      // chokidar glob patterns only work with front-slashes
+      if (!Array.isArray(files)) {
+        fileGlob = path.join(path.dirname(files), '**', '*.*').replace(/\\/g, '/');
+      } else if (files.length === 1) {
+        fileGlob = path.join(path.dirname(files[0]), '**', '*.*').replace(/\\/g, '/');
+      }
+
+      filesToWatch.push(fileGlob);
+    }
+    filesToWatch = Array.from(new Set(filesToWatch));
+
+    this.watchFiles(
+      filesToWatch,
+      function (eventName, path) {
+        self.logger.info(`${eventName}: ${path}`);
+        self.events.emit('file-' + eventName, 'asset', path);
+        self.events.emit('file-event', 'asset', path);
+      },
+      function () {
+        callback();
```

Comment:
instead of:
```
function() {
  callback();
}
```

it is possible to replace it with just `callback` ?
(there is the same pattern L96-103

Reply:
Yes, that should be possible. As mentioned above, I haven't changed any code styles in this module (for no particular reason). So this can likely be changed as well :)

Reply:
Feel free to ignore, it appears in the diff that's why I commented but this is mostly code style so minor